### PR TITLE
Settings: align NVS read calback return value to the expectation

### PR DESCRIPTION
**Pull request by nvlsianpu**
- clarify expectation by doxygen addition
- align NVS backend implementation to the expectation.

### CODE REVIEW

**Comment by nvlsianpu** [comment]
cc @kapi-no @MarekPieta

**Comment by Laczen** [comment]
I don't think this is a good fix, nvs_read will return how much can be read.

If the correct length is used from the start (it is provided to the set function) the routine will return the correct value. If a too large or too small read is done (both are possible) nvs_read will signal that something was wrong with the read command. This correction will hide the error if the requested length is larger than the available.

**Comment by nvlsianpu** [comment]
@Laczen Exact length is transferred to h_set handler so it is known for the user prior any call to the read callback. By checking this `len` a user can detect unexpected data length (simultaneously some settings data might be variable length).

I want to align this API behavior.
Otherwise need change behavior of FS and fcb back-end.

**Comment by nvlsianpu** [comment]
@Laczen ^^ So in my opinion additional signaling to the user that not all data were read doesn't make any sense.
Can you accept that?

**Comment by Laczen** [comment]
> @Laczen ^^ So in my opinion additional signaling to the user that not all data were read doesn't make any sense.
> Can you accept that?

Not really, since there is no partial read possibility, there is no real use of the length read value.

If fcb and nffs cannot return the expected read length it would be better to standardize on returning an int that signals a error in case of read error/unexpected length and 0 if read is ok.

**Comment by nvlsianpu** [comment]
@Laczen Have to go through code which uses settings in zephyr - after that I will propose an idea.

File: subsys/settings/src/settings_nvs.c [hunk 1]
```diff
@@ -32,10 +32,18 @@ static struct settings_store_itf settings_nvs_itf = {
...
 {
 	struct settings_nvs_read_fn_arg *rd_fn_arg;
+	ssize_t rc;
 
 	rd_fn_arg = (struct settings_nvs_read_fn_arg *)back_end;
 
-	return nvs_read(rd_fn_arg->fs, rd_fn_arg->id, data, len);
+	rc = nvs_read(rd_fn_arg->fs, rd_fn_arg->id, data, len);
+	if (rc > len) {
+		/* nvs_read signals that not all byte were read
```
Comment (carlescufi) [inline subsys/settings/src/settings_nvs.c]: bytes

File: include/settings/settings.h [hunk 2]
```diff
@@ -34,6 +34,19 @@ extern "C" {
  */
 #define SETTINGS_EXTRA_LEN ((SETTINGS_MAX_DIR_DEPTH - 1) + 2)
 
+/**
+ * Prototype of function used to read the data from the settings storage in
```
Comment (carlescufi) [inline include/settings/settings.h]: Remove "Prototype of"

**Review by carlescufi — CHANGES_REQUESTED** [review]

**Comment by nvlsianpu** [comment]
@carlescufi Thanks for review.  Any comment to solution?

**Comment by carlescufi** [comment]
> @carlescufi Thanks for review. Any comment to solution?

Does returning the partial length help us in any way, is it useful for the caller? Otherwise I think it would be better to return an error code, since a positive value is typically considered success, and not error.

**Comment by Laczen** [comment]
@nvlsianpu, after going through fs.h I think your proposal is ok. nvs should be modified to adhere to the standard write/read returning the length read/write.

**Review by Laczen — APPROVED** [review]

**Comment by carlescufi** [comment]
> @nvlsianpu, after going through fs.h I think your proposal is ok. nvs should be modified to adhere to the standard write/read returning the length read/written.

Great, thanks @Laczen. @nvlsianpu can you please fix the comments and I will approve?

**Comment by nvlsianpu** [comment]
fixed

**Review by carlescufi — APPROVED** [review]

**Pull request merged**